c++/src/kj/async-inl.h

```diff
@@ -1470,4 +1470,290 @@ PromiseForResult<Func, void> Executor::executeAsync(Func&& func) const {
+
+template <typename T>
+class Coroutine<T>::Adapter: private Event {
+public:
+  explicit Adapter(PromiseFulfiller<T>& f, Coroutine::Handle coroutine)
+      : fulfiller(f), coroutine(coroutine) {
+    coroutine.promise().adapter = this;
+  }
+  KJ_DISALLOW_COPY(Adapter);
+
+  ~Adapter() noexcept(false) {
+    // If a coroutine runs to completion, its frame will already have been destroyed by the time the
+    // adapter is destroyed, so this if-condition will almost always be false.
+    if (coroutine) {
+      // coroutine.promise() is a reference to our implementation object, Coroutine<T>.
+      coroutine.promise().adapter = nullptr;
+      coroutine.destroy();
+    }
+  }
+
+  void coroutineDestroyed() {
+    // The coroutine was destroyed before the adapter.
+    coroutine = nullptr;
+  }
+
+  void awaitSuspend(PromiseNode& nodeParam, ExceptionOrValue& resultParam) {
+    // Called by Awaiter::await_suspend() to tell us to wait for `node` to become ready, then store
+    // the node's result in `result`.
+
+    // We'll need to save these references for fire(). Note that Awaiter is movable, and `result` is
+    // its subobject, but it is safe to store a reference to `result` in Adapter, because by the
+    // time `await_suspend()` is called, Awaiter has already been returned from `await_transform()`
+    // and stored in its final resting place in the coroutine's frame.
+    node = &nodeParam;
+    result = &resultParam;
+
+    // We use the promise adapter as an Event, because we must be able to destroy the coroutine from
+    // within the fire() callback. The promise adapter can do this, because it lives outside of the
+    // coroutine's frame, unlike Coroutine<T> or Awaiter<U>.
+    node->onReady(this);
+  }
+
+  PromiseFulfiller<T>& fulfiller;
+  // Used to fulfill/reject the enclosing coroutine's kj::Promise<T>.
+
+private:
+  Maybe<Own<Event>> fire() override {
+    // The promise this coroutine is currently waiting on is ready. Extract its result and check for
+    // exceptions.
+
+    // Destroying or resuming the coroutine will leave these pointers dangling, so we call this
+    // function before either action.
+    auto nullifyPointers = [&]() {
+      node = nullptr;
+      result = nullptr;
+    };
+
+    node->get(*result);
+
+    // Note: the Promise::wait() implementation throws a recoverable exception if both a value and
+    // an exception are set. We only have visibility on the exception here, so the following might
+    // not have quite the same semantics.
+
+    // Check for exceptions. We could let Awaiter::await_resume() do it, but then we'd resume the
+    // coroutine just for it to rethrow, which seems pointless.
+    KJ_IF_MAYBE(exception, result->exception) {
+      nullifyPointers();
+      fulfiller.reject(kj::mv(*exception));
+      // The coroutine threw an exception. Our only choice is to destroy it.
+      //
+      // TODO(now): Should we ignore exceptions from destroy()?
```

Comment:
I'm not quite sure about this. The situation would arise with a coroutine like this:

```c++
kj::Promise<void> coro() {
  KJ_DEFER(KJ_FAIL_ASSERT("I'll throw from coroutine.destroy() (second)"));
  co_await kj::Promise<void>(KJ_EXCEPTION(FAILED, "I'll reject the returned promise (first)"));
  KJ_UNREACHABLE;
}
```

I suppose the equivalent .then() based code can't do anything particularly special to avoid the situation, and the deferred assertion would end up getting thrown.

Reply:
Actually, the equivalent `.then()` code does explicitly handle the situation where a PromiseNode both yielded an exception and threw from its destructor. The yielded exception takes priority, and the destructor exception is ignored. See `TransformPromiseNodeBase::getDepResult()`.

I think we should do the same here.

Essentially what we have here is a destructor-threw-during-unwind situation. Of course, in regular sync code, this results in std::terminate(), C++'s single worst design decision. I think, in general, these exceptions-thrown-by-destructors-during-unwind are usually "secondary" exceptions that are side effects of the "primary" exception, where the primary exception is the one that initiated the unwind. If we have to pick one to propagate, it should be the primary exception.

(Also, throwing from `fire()` will actually unwind out of the event loop, which we certainly don't want...)

Hmm... but I guess an interesting question: does `coroutine.destroy()` actually propagate exceptions from destructors? Considering that C++11 made `noexcept` the default for destructors, I could certainly imagine that they defined `destroy()` to be noexcept as well, which would be very very sad.

Reply:
Fortunately, `coroutine.destroy()` is not noexcept. Destruction semantics aren't explicitly called out, though: http://eel.is/c++draft/dcl.fct.def.coroutine#11. I'll test to see what happens if an object on the frame throws.

I suppose the code we're commenting on here won't exist soon: since we're changing to a PromiseNode implementation, I _think_ we'll be calling coroutine.destroy() exclusively from the PromiseNode's disposer, rather than here, which might change what the right answer is.

Reply:
True. But the issue might still apply on the awaiter side -- if an awaited promise resolves to an exception, and then destroying the PromiseNode also throws, you want to discard the latter exception and propagate the former.

Reply:
Coroutine destruction exceptions handled here: https://github.com/capnproto/capnproto/pull/1104/commits/0e7c329e0d815d76d4d7815a538ce8adb741400f
And awaited promise destruction exceptions handled here: https://github.com/capnproto/capnproto/pull/1104/commits/a1df89a3a276ac2c763cd0bddf9e9622df3ee531